File: kotlin/PrimePartition.kt
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Kotlin implementation of Prime Partition Algorithm
 */
object PrimePartition {

    @JvmStatic
    fun main(args: Array<String>) {
        println("=== KOTLIN VERSION ===")
        val (primes, counts) = runAlgorithm(10, listOf(1L, 2L))
        printResults(primes, counts)
    }

    private fun printResults(primes: List<Long>, counts: Map<Long, Int>) {
        val sortedPrimes = primes.sorted().distinct()
        println("Hello primes: $sortedPrimes")
        println("Total discovered: ${sortedPrimes.size}")
        println("Occurrence counts: ${counts.toSortedMap()}")
        println("Found composites: ${primes.filter { !isPrime(it) }}")
    }

    /**
     * Run algorithm for n iterations
     */
    fun runAlgorithm(iterations: Int, initial: List<Long>): Pair<List<Long>, Map<Long, Int>> {
        var current = initial
        var accPrimes = emptyList<Long>()
        var accCounts = emptyMap<Long, Int>()

        for (i in 1..iterations) {
            val found = computePrimes(current).toList()
            val distinct = found.sorted().distinct()

            // Update counts
            val newCounts = mutableMapOf<Long, Int>()
            newCounts.putAll(accCounts)
            distinct.forEach { p ->
                newCounts[p] = (newCounts[p] ?: 0) + 1
            }

            // Find next prime to add
            val currentSet = current.toSet()
            val distinctSet = found.toSet()
            val diff = distinctSet - currentSet
            val next = if (diff.isEmpty()) current else current + diff.min()

            // Update state
            accPrimes = accPrimes + distinct
            accCounts = newCounts
            current = next
        }

        return accPrimes to accCounts
    }

    /**
     * Compute all primes from the current seed set
     */
    private fun computePrimes(seeds: List<Long>): Sequence<Long> {
        if (seeds.isEmpty()) return emptySequence()

        val maxP = seeds.maxOrNull() ?: return emptySequence()
        val minR = maxP + 1
        val maxR = maxP * maxP - 1

        return sequence {
            for ((left, right) in binaryPartitions(seeds)) {
                for (exps in exponentCombinations(seeds.size, 2)) {
                    val n = seeds.size
                    val (leftExps, rightExps) = exps.splitPair(left.size)

                    val leftProd = productPow(left, leftExps)
                    val rightProd = productPow(right, rightExps)

                    for (candidate in listOf(leftProd + rightProd, abs(leftProd - rightProd))) {
                        if (candidate >= minR && candidate <= maxR && isPrime(candidate)) {
                            yield(candidate)
                        }
                    }
                }
            }
        }
    }

    /**
     * Generate all binary partitions of a list
     */
    private fun <T> binaryPartitions(list: List<T>): Sequence<Pair<List<T>, List<T>>> = sequence {
        if (list.size < 2) return@sequence

        val size = list.size
        val half = size / 2

        for (splitIdx in 1..half) {
            for (left in combinations(splitIdx, list)) {
                val remaining = list.filterNot { it in left }
                for (right in combinations(size - splitIdx, remaining)) {
                    yield(left to right)
                }
            }
        }
    }

    /**
     * Generate all k-combinations of a list
     */
    private fun <T> combinations(k: Int, list: List<T>): Sequence<List<T>> = sequence {
        if (k == 0) {
            yield(emptyList())
            return@sequence
        }
        if (list.isEmpty()) return@sequence

        val first = list.first()
        val rest = list.drop(1)

        // Include first element
        for (combo in combinations(k - 1, rest)) {
            yield(listOf(first) + combo)
        }

        // Exclude first element
        for (combo in combinations(k, rest)) {
            yield(combo)
        }
    }

    /**
     * Generate all exponent combinations
     */
    private fun exponentCombinations(n: Int, maxExp: Int): Sequence<List<Int>> = sequence {
        val exponents = (1..maxExp).toList()
        replicateM(n, exponents).forEach { yield(it) }
    }

    /**
     * Compute product with powers
     */
    private fun productPow(list: List<Long>, exps: List<Int>): Long {
        return list.zip(exps).fold(1L) { acc, (x, e) -> acc * fastPow(x, e) }
    }

    /**
     * Fast exponentiation by squaring
     */
    private fun fastPow(base: Long, exp: Int): Long {
        return when {
            exp == 0 -> 1L
            exp == 1 -> base
            exp % 2 == 0 -> {
                val half = fastPow(base, exp / 2)
                half * half
            }
            else -> base * fastPow(base, exp - 1)
        }
    }

    /**
     * Optimized primality test (6k±1 pattern)
     */
    private fun isPrime(n: Long): Boolean {
        if (n < 2) return false
        if (n == 2L || n == 3L) return true
        if (n % 2 == 0L || n % 3 == 0L) return false
        if (n < 9) return true

        val limit = isqrt(n)

        // Start at 5, check 6k-1 and 6k+1
        var k = 5L
        while (k <= limit) {
            if (n % k == 0L || n % (k + 2) == 0L) return false
            k += 6
        }

        return true
    }

    /**
     * Integer square root
     */
    private fun isqrt(n: Long): Long = sqrt(n.toDouble()).toLong()

    /**
     * Helper: replicateM (Cartesian product of n copies of a list)
     */
    private fun <T> replicateM(n: Int, list: List<T>): List<List<T>> {
        if (n == 0) return listOf(emptyList())
        if (n == 1) return list.map { listOf(it) }

        val result = mutableListOf<List<T>>()
        val firstElement = list.first()

        // Recursively build combinations
        fun recurse(remaining: Int, current: List<T>): List<List<T>> {
            if (remaining == 0) return listOf(current)

            return list.flatMap { element ->
                recurse(remaining - 1, current + element)
            }
        }

        return recurse(n, emptyList())
    }

    /**
     * Helper: Split list into two parts at index
     */
    private fun <T> List<T>.splitPair(n: Int): Pair<List<T>, List<T>> {
        return subList(0, n) to subList(n, this.size)
    }
}

/**
 * Entry point
 */
fun main() {
    PrimePartition.main(emptyArray())
}

File: julia/prime_partition.jl
#!/usr/bin/env julia

"""
Julia implementation of Prime Partition Algorithm
"""

using Printf

# ============================================================================
# Core Algorithm Functions
# ============================================================================

"""
    fast_pow(base::Int64, exp::Int)::Int64

Fast exponentiation using exponentiation by squaring.
Time complexity: O(log exp)
"""
function fast_pow(base::Int64, exp::Int)::Int64
    if exp == 0
        return 1
    elseif exp == 1
        return base
    elseif iseven(exp)
        half = fast_pow(base, exp ÷ 2)
        return half * half
    else
        return base * fast_pow(base, exp - 1)
    end
end

"""
    is_prime(n::Int64)::Bool

Optimized primality test using trial division up to √n.
Only tests odd divisors after checking 2.
Time complexity: O(√n)
"""
function is_prime(n::Int64)::Bool
    if n ≤ 1
        return false
    elseif n == 2
        return true
    elseif iseven(n)
        return false
    end
    
    sqrt_n = isqrt(n)
    for i in 3:2:sqrt_n
        if n % i == 0
            return false
        end
    end
    
    return true
end

"""
    combinations(list::Vector{T}, k::Int)::Vector{Vector{T}} where T

Generate all k-combinations of a list.
"""
function combinations(list::Vector{T}, k::Int)::Vector{Vector{T}} where T
    if k == 0
        return [T[]]
    elseif isempty(list)
        return Vector{Vector{T}}()
    end
    
    result = Vector{Vector{T}}()
    head = list[1]
    tail = list[2:end]
    
    # With head
    for combo in combinations(tail, k - 1)
        push!(result, vcat([head], combo))
    end
    
    # Without head
    append!(result, combinations(tail, k))
    
    return result
end

"""
    binary_partitions(list::Vector{T})::Vector{Tuple{Vector{T}, Vector{T}}} where T

Generate all binary partitions of a list.
Returns pairs of (left_group, right_group).
"""
function binary_partitions(list::Vector{T})::Vector{Tuple{Vector{T}, Vector{T}}} where T
    if length(list) < 2
        return Tuple{Vector{T}, Vector{T}}[]
    end
    
    result = Tuple{Vector{T}, Vector{T}}[]
    n = length(list)
    
    for i in 1:(n ÷ 2)
        for left in combinations(list, i)
            right = setdiff(list, left)
            if length(right) == n - i
                push!(result, (left, right))
            end
        end
    end
    
    return result
end

"""
    exponent_combinations(size::Int, max_exp::Int)::Vector{Vector{Int}}

Generate all combinations of exponents [1..max_exp] of given size.
"""
function exponent_combinations(size::Int, max_exp::Int)::Vector{Vector{Int}}
    if size <= 0
        return Vector{Vector{Int}}()
    end
    
    # Use Iterators.product for Cartesian product
    ranges = [1:max_exp for _ in 1:size]
    result = Vector{Vector{Int}}()
    
    for combo in Iterators.product(ranges...)
        push!(result, collect(combo))
    end
    
    return result
end

"""
    compute_primes(seeds::Vector{Int64}, max_exponent::Int=2)::Vector{Int64}

Generate and filter prime candidates from seed set using binary partitions
and powered products.
"""
function compute_primes(seeds::Vector{Int64}, max_exponent::Int=2)::Vector{Int64}
    if isempty(seeds)
        return Int64[]
    end
    
    max_prime = maximum(seeds)
    range_start = max_prime + 1
    range_end = max_prime^2 - 1
    
    candidates = Int64[]
    
    for (left, right) in binary_partitions(seeds)
        for exps in exponent_combinations(length(seeds), max_exponent)
            left_exps = exps[1:length(left)]
            right_exps = exps[length(left)+1:end]
            
            # Compute powered products
            left_prod = prod(fast_pow(left[i], left_exps[i]) for i in 1:length(left))
            right_prod = prod(fast_pow(right[i], right_exps[i]) for i in 1:length(right))
            
            # Generate candidates
            push!(candidates, left_prod + right_prod)
            push!(candidates, abs(left_prod - right_prod))
        end
    end
    
    # Filter and deduplicate
    filtered = filter(c -> range_start ≤ c ≤ range_end, candidates)
    unique_candidates = unique(filtered)
    primes = filter(is_prime, unique_candidates)
    
    return sort(primes)
end

"""
    run_algorithm(iterations::Int, initial::Vector{Int64}, max_exponent::Int=2)

Run the partition algorithm for specified iterations.
Returns tuple of (all_primes, occurrence_counts).
"""
function run_algorithm(iterations::Int, initial::Vector{Int64}, max_exponent::Int=2)
    current = copy(initial)
    acc_primes = Int64[]
    acc_counts = Dict{Int64, Int}()
    
    for iter in 1:iterations
        found = compute_primes(current, max_exponent)
        distinct = unique(found)
        
        # Update occurrence counts
        for prime in found
            acc_counts[prime] = get(acc_counts, prime, 0) + 1
        end
        
        # Accumulate primes
        append!(acc_primes, distinct)
        
        # Find next prime to add
        new_primes = setdiff(distinct, current)
        if !isempty(new_primes)
            min_new = minimum(new_primes)
            push!(current, min_new)
        end
    end
    
    return (acc_primes, acc_counts)
end

# ============================================================================
# Main Execution
# ============================================================================

function main()
    println("=== JULIA VERSION ===")
    
    # Run algorithm
    (primes, counts) = run_algorithm(10, Int64[1, 2], 2)
    
    # Get unique sorted primes
    sorted_primes = sort(unique(primes))
    
    # Display results
    println("Hello primes: ", sorted_primes)
    println("Total discovered: ", length(sorted_primes))
    
    # Check for composites (should be none)
    composites = filter(p -> !is_prime(p), primes)
    println("Found composites: ", composites)
    
    # Optional: Show occurrence counts
    # println("\nOccurrence counts:")
    # for (prime, count) in sort(collect(counts))
    #     println("  $prime: $count")
    # end
end

# Run if executed as script
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
